File: cathode/src/main/java/net/simonvt/cathode/ui/lists/ListsFragment.java
/*
 * Copyright (C) 2016 Simon Vig Therkildsen
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package net.simonvt.cathode.ui.lists;

import android.app.Activity;
import android.os.Bundle;
import android.view.MenuItem;
import androidx.appcompat.widget.Toolbar;
import androidx.lifecycle.Observer;
import androidx.lifecycle.ViewModelProviders;
import dagger.android.support.AndroidSupportInjection;
import java.util.List;
import javax.inject.Inject;
import net.simonvt.cathode.R;
import net.simonvt.cathode.common.entity.UserList;
import net.simonvt.cathode.common.ui.fragment.ToolbarSwipeRefreshRecyclerFragment;
import net.simonvt.cathode.jobqueue.Job;
import net.simonvt.cathode.jobqueue.JobManager;
import net.simonvt.cathode.remote.sync.lists.SyncLists;
import net.simonvt.cathode.ui.ListNavigationListener;

public class ListsFragment extends ToolbarSwipeRefreshRecyclerFragment<ListsAdapter.ViewHolder>
    implements ListsAdapter.OnListClickListener {

  public static final String TAG = "net.simonvt.cathode.ui.lists.ListsFragment";

  static final String DIALOG_LIST_CREATE = "net.simonvt.cathode.ui.HomeActivity.createListFragment";

  @Inject JobManager jobManager;

  private ListNavigationListener listener;

  private ListsViewModel viewModel;

  private ListsAdapter adapter;

  @Override public void onAttach(Activity activity) {
    super.onAttach(activity);
    listener = (ListNavigationListener) activity;
  }

  @Override public void onCreate(Bundle inState) {
    super.onCreate(inState);
    AndroidSupportInjection.inject(this);
    setTitle(R.string.navigation_lists);
    setEmptyText(R.string.empty_lists);

    viewModel = ViewModelProviders.of(this).get(ListsViewModel.class);
    viewModel.getLists().observe(this, new Observer<List<UserList>>() {
      @Override public void onChanged(List<UserList> userLists) {
        setLists(userLists);
      }
    });
  }

  @Override protected int getColumnCount() {
    return 1;
  }

  private Job.OnDoneListener onDoneListener = new Job.OnDoneListener() {
    @Override public void onDone(Job job) {
      setRefreshing(false);
    }
  };

  @Override public void onRefresh() {
    Job job = new SyncLists();
    job.registerOnDoneListener(onDoneListener);
    jobManager.addJob(job);
  }

  @Override public void onListClicked(long listId, String listName) {
    listener.onShowList(listId, listName);
  }

  @Override public boolean displaysMenuIcon() {
    return true;
  }

  @Override public void createMenu(Toolbar toolbar) {
    super.createMenu(toolbar);
    toolbar.inflateMenu(R.menu.fragment_lists);
  }

  @Override public boolean onMenuItemClick(MenuItem item) {
    switch (item.getItemId()) {
      case R.id.menu_list_create:
        CreateListFragment dialog = new CreateListFragment();
        dialog.show(getFragmentManager(), DIALOG_LIST_CREATE);
        return true;
    }

    return super.onMenuItemClick(item);
  }

  private void setLists(List<UserList> userLists) {
    if (adapter == null) {
      adapter = new ListsAdapter(this, requireContext());
      setAdapter(adapter);
    }

    adapter.setList(userLists);
  }
}

File: cathode/src/main/java/net/simonvt/cathode/ui/lists/ListsFragment.kt
/*
 * Copyright (C) 2016 Simon Vig Therkildsen
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package net.simonvt.cathode.ui.lists

import android.content.Context
import android.os.Bundle
import android.view.MenuItem
import android.view.View
import androidx.appcompat.widget.Toolbar
import androidx.fragment.app.viewModels
import androidx.lifecycle.Observer
import net.simonvt.cathode.R
import net.simonvt.cathode.common.ui.fragment.ToolbarSwipeRefreshRecyclerFragment
import net.simonvt.cathode.common.ui.instantiate
import net.simonvt.cathode.entity.UserList
import net.simonvt.cathode.settings.TraktLinkSettings
import net.simonvt.cathode.ui.CathodeViewModelFactory
import net.simonvt.cathode.ui.ListNavigationListener
import javax.inject.Inject

class ListsFragment @Inject constructor(
  private val viewModelFactory: CathodeViewModelFactory
) : ToolbarSwipeRefreshRecyclerFragment<ListsAdapter.ViewHolder>(),
  ListsAdapter.OnListClickListener {

  private lateinit var listener: ListNavigationListener

  private val viewModel: ListsViewModel by viewModels { viewModelFactory }

  private var adapter: ListsAdapter? = null

  override fun onAttach(context: Context) {
    super.onAttach(context)
    listener = requireActivity() as ListNavigationListener
  }

  override fun onCreate(inState: Bundle?) {
    super.onCreate(inState)
    setTitle(R.string.navigation_lists)
    setEmptyText(R.string.empty_lists)

    viewModel.loading.observe(this, Observer { loading -> setRefreshing(loading) })
    viewModel.lists.observe(this, Observer { userLists -> setLists(userLists) })
  }

  override fun onViewCreated(view: View, inState: Bundle?) {
    super.onViewCreated(view, inState)
    swipeRefreshLayout.isEnabled = TraktLinkSettings.isLinked(requireContext())
  }

  override fun getColumnCount(): Int {
    return 1
  }

  override fun onRefresh() {
    viewModel.refresh()
  }

  override fun onListClicked(listId: Long, listName: String) {
    listener.onShowList(listId, listName)
  }

  override fun displaysMenuIcon(): Boolean {
    return true
  }

  override fun createMenu(toolbar: Toolbar) {
    super.createMenu(toolbar)
    toolbar.inflateMenu(R.menu.fragment_lists)
  }

  override fun onMenuItemClick(item: MenuItem): Boolean {
    when (item.itemId) {
      R.id.menu_list_create -> {
        parentFragmentManager.instantiate(CreateListFragment::class.java)
          .show(parentFragmentManager, DIALOG_LIST_CREATE)
        return true
      }
    }

    return super.onMenuItemClick(item)
  }

  private fun setLists(userLists: List<UserList>) {
    if (adapter == null) {
      adapter = ListsAdapter(this, requireContext())
      setAdapter(adapter)
    }

    adapter!!.setList(userLists)
  }

  companion object {

    const val TAG = "net.simonvt.cathode.ui.lists.ListsFragment"

    private const val DIALOG_LIST_CREATE = "net.simonvt.cathode.ui.HomeActivity.createListFragment"
  }
}
